### PR TITLE
deployment: Add NomadDeployer

### PR DESCRIPTION
**Pull request by nikochiko**

### CODE REVIEW

File: capstone/config.py [hunk 1]
```diff
@@ -7,11 +7,13 @@
 
 capstone_api_token = os.getenv("CAPSTONE_API_TOKEN", "test123")
 
-gitto_base_url = os.getenv("CAPSTONE_GITTO_BASE_URL", "http://localhost:8080")
+gitto_base_url = os.getenv("CAPSTONE_GITTO_BASE_URL", "http://localhost:7878")
 gitto_api_token = os.getenv("CAPSTONE_GITTO_API_TOKEN", "gitto")
 
 data_dir = os.getenv("CAPSTONE_DATA_DIR", "data")
 
+docker_registry = os.getenv("CAPSTONE_DOCKER_REGISTRY", "registry.local.pipal.in:8080")
```
Comment (nikochiko) [inline capstone/config.py:15]: This shouldn't run on port 8080.

File: capstone/deployment.py [hunk 2]
```diff
@@ -101,3 +111,137 @@ def add_sys_path(path):
...
+        self.git_url = git_url
+        self.cwd = None
+
+    def chdir(self, directory):
+        self.logger.info("")
+        self.logger.info("$ cd %s", directory)
+        self.cwd = os.path.join(self.cwd or "", directory)
+
+    def run(self):
+        self.logger.info(f"Starting DeploTask {self.task_id} to deploy {self.hostname}")
```
Comment (nikochiko) [inline capstone/deployment.py:197]: `s/DeploTask/DeployTask`

File: capstone/deployment.py [hunk 3]
```diff
@@ -101,3 +111,137 @@ def add_sys_path(path):
...
+        logfile = log_root.joinpath("tasks", site, filename)
+        logfile.parent.mkdir(parents=True, exist_ok=True)
+        handler = logging.FileHandler(str(logfile))
+        formatter = logging.Formatter(
+            fmt=f'[%(asctime)s] [%(levelname)s] %(message)s',
+            datefmt='%Y-%m-%d %H:%M:%S'
+        )
+        handler.setFormatter(formatter)
+        logger.addHandler(handler)
+        logger.addHandler(logging.StreamHandler(sys.stdout))
```
Comment (nikochiko) [inline capstone/deployment.py:180]: Why is this logging to stdout and not stderr?

File: nomad/jobs/nginx.nomad.hcl [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        volumes = [
+          "local:/etc/nginx/conf.d",
+        ]
+      }
+
+      template {
+        data = <<EOF
+{{ range services }}
+
+{{ $host := "localhost" }}
```
Comment (nikochiko) [inline nomad/jobs/nginx.nomad.hcl:35]: Why is $host being set to localhost here and then to .ServiceMeta.host later?

File: capstone/deployment.py [hunk 5]
```diff
@@ -101,3 +111,137 @@ def add_sys_path(path):
...
+
+    task "{{name}}" {
+
+      driver = "docker"
+
+      config {
+        image = "{{docker_image}}"
+        force_pull = true
+        ports = ["web"]
+      }
```
Comment (nikochiko) [inline capstone/deployment.py:147]: where do we specify the local docker registry to pull this image from?

File: capstone/config.py (hunk 1 above)
Comment (anandology) [reply capstone/config.py:15]: why? It is running in Nomad. Nginx in Nomad is running on port 8080.

File: capstone/deployment.py (hunk 3 above)
Comment (anandology) [reply capstone/deployment.py:180]: no real preference. Using stderr was messing up with some earlier prints, so started using stdout. I'm okay to move to stderr.

File: nomad/jobs/nginx.nomad.hcl (hunk 4 above)
Comment (anandology) [reply nomad/jobs/nginx.nomad.hcl:35]: `.ServiceMeta.host` is available for each service in the inner loop, not here. So I'm initializing it to `localhost` and updating it in inner loop.

File: capstone/deployment.py (hunk 5 above)
Comment (anandology) [reply capstone/deployment.py:147]: docker image name is generated. See `docker_registry` in `config.py`.

**Pull request merged**